extensions/azdata/src/azdata.ts

```diff
@@ -218,23 +220,112 @@ export async function upgradeAzdata(): Promise<void> {
+ */
+export async function checkAndInstallAzdata(userRequested: boolean = false): Promise<IAzdataTool | undefined> {
+	try {
+		const azdata = await findAzdata(); // find currently installed Azdata
+		// Don't block on this since we want the extension to finish activating without needing user input
+		checkAndUpgradeAzdata(azdata, userRequested).catch(err => {
+			//update if one is available and user wants it.
+			vscode.window.showWarningMessage(loc.updateError(err));
+			Logger.log(loc.updateError(err));
+		}).then(() => {
+			return findAzdata(); // now again find and return the currently installed azdata
+		});
+	} catch (err) {
+		// Don't block on this since we want the extension to finish activating without needing user input.
+		// Calls will be made to handle azdata not being installed
+		await promptToInstallAzdata(userRequested).catch(e => console.log(`Unexpected error prompting to install azdata ${e}`));
+	}
+	return undefined;
+}
+
+/**
+ * Checks whether a newer version of azdata is available - and if it is then invokes the process of azdata upgrade.
+ * @param currentAzdata The current version of azdata to check again
  * Checks whether a newer version of azdata is available - and if it is prompts the user to download and
  * install it.
  * @param currentAzdata The current version of azdata to check against
+ * @param userRequested true means that this operation by was requested by a user by executing an ads command.
  */
-export async function checkAndUpgradeAzdata(currentAzdata?: IAzdataTool): Promise<void> {
+export async function checkAndUpgradeAzdata(currentAzdata?: IAzdataTool, userRequested: boolean = false): Promise<void> {
 	if (currentAzdata === undefined) {
 		currentAzdata = await findAzdata();
 	}
 	const newVersion = await discoverLatestAvailableAzdataVersion();
 	if (newVersion.compare(currentAzdata.cachedVersion) === 1) {
-		const response = await vscode.window.showInformationMessage(loc.promptForAzdataUpgrade(newVersion.raw), loc.yes, loc.no);
-		if (response === loc.yes) {
-			await upgradeAzdata();
+		Logger.log(loc.foundAzdataVersionToUpgradeTo(newVersion.raw, currentAzdata.cachedVersion.raw));
+		await promptToUpgradeAzdata(userRequested).catch(e => console.log(`Unexpected error prompting to upgrade azdata ${e}`));
+	}
+}
+
+async function promptToInstallAzdata(userRequested: boolean = false): Promise<void> {
+	let response: string | undefined = loc.yes;
+	const config = <AzdataDeployOption>getConfig(azdataAutoInstallKey);
+	Logger.log(loc.autoDeployConfig(azdataAutoInstallKey, config));
+	if (userRequested) {
+		Logger.log(loc.userRequestedInstall);
+	}
+	if (config === AzdataDeployOption.never && !userRequested) {
+		Logger.log(loc.skipInstall(config));
+		return;
+	}
+	if (config === AzdataDeployOption.userPrompt) {
+		response = await vscode.window.showErrorMessage(loc.couldNotFindAzdataWithPrompt, loc.yes, loc.no, loc.always, loc.never);
```

Comment:
I don't think "always" really makes sense here. It's not going to be a normal thing to prompt users for installation so having it prompt each time (or never if they ignore) should be enough.

Reply:
removed.